xla/stream_executor/sycl/sycl_gpu_runtime.cc

```diff
@@ -107,4 +196,514 @@ absl::StatusOr<::sycl::device> SyclDevicePool::GetDevice(int device_ordinal) {
+  bool async = (dst_alloc_type == ::sycl::usm::alloc::host);
+  return MemcpyDeviceToHost(stream_handle, dst_host, src_device, byte_count,
+                            async);
+}
+
+absl::Status SyclMemcpyHostToDeviceAsync(::sycl::queue* stream_handle,
+                                         void* dst_device, const void* src_host,
+                                         size_t byte_count) {
+  if (dst_device == nullptr || src_host == nullptr) {
+    return absl::InvalidArgumentError(
+        "SyclMemcpyHostToDeviceAsync: Null pointer provided for destination or "
+        "source.");
+  }
+  if (stream_handle == nullptr) {
+    return absl::InvalidArgumentError(
+        "SyclMemcpyHostToDeviceAsync: Null stream handle provided.");
+  }
+  if (byte_count == 0) {
+    LOG(WARNING)
+        << "SyclMemcpyHostToDeviceAsync: Attempting to copy zero bytes, "
+           "skipping operation.";
+    return absl::OkStatus();
+  }
+  ::sycl::usm::alloc src_alloc_type =
+      get_pointer_type(src_host, stream_handle->get_context());
+  bool async = (src_alloc_type == ::sycl::usm::alloc::host);
+  return MemcpyHostToDevice(stream_handle, dst_device, src_host, byte_count,
+                            async);
+}
+
+absl::Status SyclMemcpyDeviceToDeviceAsync(::sycl::queue* stream_handle,
+                                           void* dst_device,
+                                           const void* src_device,
+                                           size_t byte_count) {
+  if (dst_device == nullptr || src_device == nullptr) {
+    return absl::InvalidArgumentError(
+        "SyclMemcpyDeviceToDeviceAsync: Null pointer provided for destination "
+        "or source.");
+  }
+  if (stream_handle == nullptr) {
+    return absl::InvalidArgumentError(
+        "SyclMemcpyDeviceToDeviceAsync: Null stream handle provided.");
+  }
+  if (byte_count == 0) {
+    LOG(WARNING)
+        << "SyclMemcpyDeviceToDeviceAsync: Attempting to copy zero bytes, "
+           "skipping operation.";
+    return absl::OkStatus();
+  }
+  return MemcpyDeviceToDevice(stream_handle, dst_device, src_device, byte_count,
+                              /*async=*/true);
+}
+
+absl::Status SyclMemsetDevice(int device_ordinal, void* dst_device,
+                              unsigned char value, size_t count) {
+  if (dst_device == nullptr) {
+    return absl::InvalidArgumentError(
+        "SyclMemsetDevice: Null pointer provided for destination.");
+  }
+  if (count == 0) {
+    LOG(WARNING) << "SyclMemsetDevice: Attempting to set zero bytes, "
+                    "skipping operation.";
+    return absl::OkStatus();
+  }
+  TF_RETURN_IF_ERROR(IsValidDeviceOrdinal(device_ordinal, "SyclMemsetDevice"));
+  TF_ASSIGN_OR_RETURN(StreamPtr stream_handle,
+                      SyclStreamPool::GetDefaultStream(device_ordinal));
+  return MemsetDevice(stream_handle.get(), dst_device, value, count);
+}
+
+absl::Status SyclMemsetDeviceAsync(::sycl::queue* stream_handle,
```

Comment:
Can you either make `stream_handle` a reference instead of a pointer (preferred) or add the `absl_nonnull` attribute?

Reply:
Done.